### PR TITLE
De-optimize txs requesting to handle corner cases properly

### PR DESCRIPTION
**Pull request by flcl42**
Fixes hive tests

- Tx still should be requested if correct and incorrect txs are being returned by different peers
- If a peer does not send tx back, it will probably block from requesting from others In tx pool presence check is enough

_Please choose one of the keywords above to refer to the issue this PR solves followed by the issue number (e.g. Fixes #000). If no issue number, remove the line. Also, remove everything marked optional that is not applicable. Remove this note after reading._

## Changes

- Remove cache for requesting tx hashes

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No (covered by hive)

### CODE REVIEW

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/PooledTxsRequestor.cs [hunk 1]
```diff
@@ -25,8 +25,6 @@ public class PooledTxsRequestor(ITxPool txPool, ITxPoolConfig txPoolConfig, ISpe
             ? long.MaxValue
             : txPoolConfig.MaxBlobTxSize.Value + (long)specProvider.GetFinalMaxBlobGasPerBlock();
 
-        private readonly ClockKeyCache<ValueHash256> _pendingHashes = new(MemoryAllowance.TxHashCacheSize);
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/PooledTxsRequestor.cs:28]: Just shrink the cache so churns faster?

Then will act to throttle parallel requests; but not so big that they get lost if peer doesn't respond?

Comment (flcl42) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/PooledTxsRequestor.cs:28]: Remade it so know it requests from all the peers if initial one failed to respond properly and in time

File: src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    where TNodeId : notnull, IEquatable<TNodeId>
+{
+    AnnounceResult Announced(TResourceId resourceId, TNodeId nodeId, Action request);
+    void Received(TResourceId resourceId);
+}
+public abstract class SimpleRetryCache
+{
+    public const int TimeoutMs = 2500;
+    public const int CheckMs = 300;
+}
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs]: I think not needed for consts

File: src/Nethermind/Nethermind.TxPool/Filters/PriorityFeeTooLowFilter.cs [hunk 3]
```diff
@@ -12,7 +12,7 @@ namespace Nethermind.TxPool.Filters;
 public class PriorityFeeTooLowFilter : IIncomingTxFilter
 {
     private readonly ILogger _logger;
-    private static readonly UInt256 _minBlobsPriorityFee = 1.GWei();
+    private static readonly UInt256 _minBlobsPriorityFee = 0.GWei();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/Filters/PriorityFeeTooLowFilter.cs]: Why are we reducing it to 0?

File: src/Nethermind/Nethermind.TxPool/ITxPool.cs [hunk 4]
```diff
@@ -60,5 +60,7 @@ bool TryGetBlobAndProofV1(byte[] blobVersionedHash,
         public bool AcceptTxWhenNotSynced { get; set; }
         bool SupportsBlobs { get; }
         long PendingTransactionsAdded { get; }
+
+        ISimpleRetryCache<ValueHash256, Guid> RetryCache { get; }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/ITxPool.cs]: Instead of it living in `ITxPool`, this should be injected into `TxPool` and `PooledTxRequestor` by DI.

File: src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+public class SimpleRetryCache<TResourceId, TNodeId> : SimpleRetryCache, ISimpleRetryCache<TResourceId, TNodeId>
+    where TResourceId : struct, IEquatable<TResourceId>
+    where TNodeId : notnull, IEquatable<TNodeId>
+{
+    private readonly ConcurrentDictionary<TResourceId, Dictionary<TNodeId, Action>> _retryRequests = new();
+    private readonly ConcurrentQueue<(TResourceId ResourceId, DateTimeOffset Expires)> expiringQueue = new();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs]:
```suggestion
    private readonly ConcurrentQueue<(TResourceId ResourceId, DateTimeOffset Expires)> _expiringQueue = new();
```

File: src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            return [];
+        }
+
+        Dictionary<TNodeId, Action> Update(TNodeId nodeId, TResourceId resourceId, Action request, Dictionary<TNodeId, Action> dictionary)
+        {
+            if (_logger.IsTrace) _logger.Trace($"Announced {resourceId} by {nodeId}: UPDATE");
+
+            dictionary[nodeId] = request;
+            return dictionary;
+        }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs]: Maybe it is me, but might be easier to read if you actually inline those sub-methods as simple lambda's

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/PooledTxsRequestor.cs [hunk 7]
```diff
@@ -94,30 +90,69 @@ public void RequestTransactionsEth68(Action<V66.Messages.GetPooledTransactionsMe
...
+            for (int i = 0; i < hashes.Length; i++)
+            {
+                Hash256 hash = hashes[i];
+                if (!txPool.IsKnown(hash))
+                {
+                    if (txPool.RetryCache.Announced(hash, sessionId, () =>
+                    {
+                        ArrayPoolList<Hash256> hashesToRetry = new(1) { hash };
+                        GetPooledTransactionsMessage msg65 = new(hashesToRetry);
+                        send(msg65);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/PooledTxsRequestor.cs]: We could avoid allocations of delegates, if we passed 2 separate actions. One for constructing the message and one for Send.

A bit less clean but no delegates allocated as the constructor one can be saved in private field (or be a method)

**Review by LukaszRozmej — COMMENTED** [review]
Feels a bit overcomplicated.

File: src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs (hunk 5 above)
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs]: Maybe instead of tracking `_retryRequests` internally the `RetryCache` could just sound an event that retry is needed and it could be handled by `PooledTxsRequestor` which would actually retry? Would make it less complicated and removed the need for passing `Action`s. Made the separation of concerns better.

File: src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs (hunk 2 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs]: It's for tests, to be able to access `SimpleRetryCache.TimeoutMs` without generic args. Let me know it's still too ugly to keep considering this argument

File: src/Nethermind/Nethermind.TxPool/Filters/PriorityFeeTooLowFilter.cs (hunk 3 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.TxPool/Filters/PriorityFeeTooLowFilter.cs]: Should not be changed

File: src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs (hunk 6 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs]: 👍

File: src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs (hunk 5 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs]: 👍

File: src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs]: IMO Not worth it to be separate class

File: src/Nethermind/Nethermind.TxPool/ITxPool.cs (hunk 4 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.TxPool/ITxPool.cs]: Hid it fully in txpool. The cool part of it is that new method is extendible to easily track tx sizes and types. That we don't yet validate but will do so soon

File: src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs (hunk 5 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs]: I'd remove actions fully in all the handlers, passing sessions instead

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/PooledTxsRequestor.cs (hunk 7 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/PooledTxsRequestor.cs]: Let's just get rid of send later?

File: src/Nethermind/Nethermind.Network.Contract/Messages/IResourceRequestMessage.cs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+namespace Nethermind.Network.Contract.Messages;
+
+public interface IResourceRequestMessage<TMessage, TResourceId>
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network.Contract/Messages/IResourceRequestMessage.cs]:
```suggestion
public interface IResourceRequestMessage<out TMessage, in TResourceId>
```

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/Eth65ProtocolHandler.cs [hunk 9]
```diff
@@ -70,16 +71,16 @@ public override void HandleMessage(ZeroPacket message)
...
                     BackgroundTaskScheduler.ScheduleBackgroundTask(getPooledTxMsg, Handle);
                     break;
-                case Eth65MessageCode.NewPooledTransactionHashes:
+                case Eth65MessageCode.PooledTransactions:
                     if (CanReceiveTransactions)
                     {
-                        NewPooledTransactionHashesMessage newPooledTxMsg = Deserialize<NewPooledTransactionHashesMessage>(message.Content);
-                        ReportIn(newPooledTxMsg, size);
-                        Handle(newPooledTxMsg);
+                        PooledTransactionsMessage pooledTxMsg = Deserialize<PooledTransactionsMessage>(message.Content);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/Eth65ProtocolHandler.cs:78]:
```suggestion
                        using PooledTransactionsMessage pooledTxMsg = Deserialize<PooledTransactionsMessage>(message.Content);
```

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/Eth65ProtocolHandler.cs [hunk 10]
```diff
@@ -182,10 +185,40 @@ protected override void SendNewTransactionsCore(IEnumerable<Transaction> txs, bo
...
         }
 
-        private void SendNewPooledTransactionMessage(IOwnedReadOnlyList<Hash256> hashes)
+        protected virtual void RequestPooledTransactions(IOwnedReadOnlyList<Hash256> hashes)
+        {
+            ArrayPoolList<Hash256> newTxHashes = AddMarkUnknownHashes(hashes.AsSpan());
+
+            if (newTxHashes.Count == 0)
+            {
+                return;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/Eth65ProtocolHandler.cs:192]:
```suggestion
                newTxHashes.Dispose();
                return;
```

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V66/Eth66ProtocolHandler.cs [hunk 11]
```diff
@@ -196,13 +192,43 @@ protected override void Handle(NewPooledTransactionHashesMessage msg)
...
+        protected override void RequestPooledTransactions(IOwnedReadOnlyList<Hash256> hashes)
+        {
+            void RequestPooledTransactionsEth66(IOwnedReadOnlyList<Hash256> hashes) => Send(new GetPooledTransactionsMessage(hashes));
+
+            ArrayPoolList<Hash256> discoveredTxHashes = AddMarkUnknownHashes(hashes.AsSpan());
+
+            if (discoveredTxHashes.Count == 0)
+            {
+                return;
+            }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V66/Eth66ProtocolHandler.cs]:
```suggestion
            if (discoveredTxHashes.Count == 0)
            {            
                discoveredTxHashes.Dispose();
                return;
            }
```

Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V66/Eth66ProtocolHandler.cs]: Can we share code for that with base?

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V68/Eth68ProtocolHandler.cs [hunk 12]
```diff
@@ -96,11 +96,74 @@ private void Handle(NewPooledTransactionHashesMessage68 msg)
...
 
+    public void RequestTransactions(IOwnedReadOnlyList<Hash256> hashes, IOwnedReadOnlyList<int> sizes, IOwnedReadOnlyList<byte> types)
+    {
+        using ArrayPoolList<(Hash256 Hash, byte Type, int Size)> discoveredTxHashesAndSizes = AddMarkUnknownHashes(hashes.AsSpan(), sizes.AsSpan(), types.AsSpan());
+        if (discoveredTxHashesAndSizes.Count == 0)
+        {
+            hashes.Dispose();
+            sizes.Dispose();
+            types.Dispose();
+            return;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V68/Eth68ProtocolHandler.cs:114]:
```suggestion
            hashes.Dispose();
            sizes.Dispose();
            types.Dispose();
            discoveredTxHashesAndSizes.Dispose();
            return;
```

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V68/Eth68ProtocolHandler.cs [hunk 13]
```diff
@@ -96,11 +96,74 @@ private void Handle(NewPooledTransactionHashesMessage68 msg)
...
+        {
+            Hash256 hash = hashes[i];
+            if (!_txPool.IsKnown(hash) && !_txPool.ContainsTx(hash, (TxType)types[i]))
+            {
+                if (_txPool.AnnounceTx(hash, this) is AnnounceResult.New)
+                {
+                    discoveredTxHashesAndSizes.Add((hash, types[i], sizes[i]));
+                }
+            }
+        }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V68/Eth68ProtocolHandler.cs]: Seems to me this could be somewhat shared with generics with base code?

File: src/Nethermind/Nethermind.Runner/packages.lock.json [hunk 14]
```diff
@@ -1835,6 +1835,7 @@
           "Nethermind.Crypto": "[1.35.0-unstable, )",
           "Nethermind.Db": "[1.35.0-unstable, )",
           "Nethermind.Evm": "[1.35.0-unstable, )",
+          "Nethermind.Network.Contract": "[1.35.0-unstable, )",
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Runner/packages.lock.json:1839]: Explain?

File: src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+public class SimpleRetryCache<TMessage, TResourceId>
+    where TMessage : IResourceRequestMessage<TMessage, TResourceId>
+    where TResourceId : struct, IEquatable<TResourceId>
+{
+    private readonly int _timeoutMs;
+    private readonly int _checkMs;
+    private readonly int _requestingCacheSize;
+
+    private readonly ConcurrentDictionary<TResourceId, HashSet<IMessageHandler<TMessage>>> _retryRequests = new();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs]: Idea:

Use ArrayPoolList instead of HashSet to not allocate much (We could create ArrayPoolSet but that is harder).

File: src/Nethermind/Nethermind.Network.Test/P2P/Subprotocols/Eth/V66/PooledTransactionsRequestingTests.cs [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+[TestFixture, Parallelizable(ParallelScope.Self)]
+public class PooledTransactionsRequestingTests
+{
+    private ISession _session = null!;
+    private ISession _session2 = null!;
+    private Eth66ProtocolHandler _handler = null!;
+    private ArrayPoolList<Transaction> _txs = null!;
+    private IMessageSerializationService _svc = null!;
+    private Block _genesisBlock = null!;
+    private CompositeDisposable _disposables = null!;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network.Test/P2P/Subprotocols/Eth/V66/PooledTransactionsRequestingTests.cs:53]: Relying on fields make it not fully parallelizable with each other. If they are fast that's probably fine.

File: src/Nethermind/Nethermind.Runner/packages.lock.json (hunk 14 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Runner/packages.lock.json:1839]: There is a project we have that is suitable for code contracts like those messages. used also because it solves a dependency loop

File: src/Nethermind/Nethermind.Network.Test/P2P/Subprotocols/Eth/V66/PooledTransactionsRequestingTests.cs (hunk 16 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Network.Test/P2P/Subprotocols/Eth/V66/PooledTransactionsRequestingTests.cs:53]: Added FixtureLifeCycle(LifeCycle.InstancePerTestCase), it's indeed needs time to run, but would be so nice to not run it one by one

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/Eth65ProtocolHandler.cs (hunk 9 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/Eth65ProtocolHandler.cs:78]: We will use transactions from it later in a scheduled task, so can't dispose for now

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V62/Eth62ProtocolHandler.cs [hunk 17]
```diff
@@ -240,7 +237,7 @@ private void Handle(StatusMessage status)
         protected void Handle(TransactionsMessage msg)
         {
             IOwnedReadOnlyList<Transaction> iList = msg.Transactions;
-            BackgroundTaskScheduler.ScheduleBackgroundTask((iList, 0), _handleSlow);
+            BackgroundTaskScheduler.ScheduleBackgroundTask((iList, 0), HandleSlow);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V62/Eth62ProtocolHandler.cs]: Does it allocate on each call? Not sure if they already changed that in dotnet.

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/Eth65ProtocolHandler.cs [hunk 18]
```diff
@@ -180,10 +176,72 @@ protected override void SendNewTransactionsCore(IEnumerable<Transaction> txs, bo
...
+            }
+            else
+            {
+                try
+                {
+                    for (int start = 0; start < newTxHashes.Count; start += MaxNumberOfTxsInOneMsg)
+                    {
+                        int end = Math.Min(start + MaxNumberOfTxsInOneMsg, newTxHashes.Count);
+
+                        ArrayPoolList<Hash256> hashesToRequest = new(end - start);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/Eth65ProtocolHandler.cs:207]:
```suggestion
                        ArrayPoolList<Hash256> hashesToRequest = new(end - start + 1);
```

Comment (flcl42) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/Eth65ProtocolHandler.cs:207]: it seems like we don't really needs this +1:
> int end = Math.Min(start + MaxNumberOfTxsInOneMsg, newTxHashes.Count);
so end - start is either MaxNumberOfTxsInOneMsg or newTxHashes.Count, which looks ok

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V68/Eth68ProtocolHandler.cs (hunk 13 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V68/Eth68ProtocolHandler.cs]: Done. eth68 has several small diffs, so only 65th and 66th code parts are shared

File: src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs (hunk 15 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.TxPool/SimpleRetryCache.cs]: Added PooledSet, wdyt?

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V62/Eth62ProtocolHandler.cs (hunk 17 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V62/Eth62ProtocolHandler.cs]: Still allocates, rolled back

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V68/Eth68ProtocolHandler.cs [hunk 19]
```diff
@@ -94,11 +94,86 @@ private void Handle(NewPooledTransactionHashesMessage68 msg)
...
+    protected void RequestPooledTransactions(IOwnedReadOnlyList<Hash256> hashes, IOwnedReadOnlyList<int> sizes, IOwnedReadOnlyList<byte> types)
+    {
+        using ArrayPoolList<int> newTxHashesIndexes = AddMarkUnknownHashes(hashes.AsSpan());
+
+        if (newTxHashesIndexes.Count == 0)
+        {
+            hashes.Dispose();
+            sizes.Dispose();
+            types.Dispose();
+            newTxHashesIndexes.Dispose();
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V68/Eth68ProtocolHandler.cs]: This one is in `using` so don't need to `.Dispose`

**Review by benaadams — APPROVED** [review]

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V68/Eth68ProtocolHandler.cs (hunk 19 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V68/Eth68ProtocolHandler.cs]: Indeed

File: src/Nethermind/Nethermind.Network.Test/ProtocolsManagerTests.cs [hunk 20]
```diff
@@ -114,12 +109,12 @@ public Context()
             _peerStorage = Substitute.For<INetworkStorage>();
             _syncPeerPool = Substitute.For<ISyncPeerPool>();
             _gossipPolicy = Substitute.For<IGossipPolicy>();
+            _gossipPolicy = Substitute.For<IGossipPolicy>();
```
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.Network.Test/ProtocolsManagerTests.cs]: Two identical lines `_gossipPolicy = Substitute.For<IGossipPolicy>();`

File: src/Nethermind/Nethermind.TxPool/TxPool.cs [hunk 21]
```diff
@@ -542,6 +547,11 @@ public AcceptTxResult SubmitTx(Transaction tx, TxHandlingOptions handlingOptions
                     _transactionSnapshot = null;
             }
 
+            if (accepted)
+            {
+                _retryCache.Received(tx.Hash!);
```
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.TxPool/TxPool.cs]: Shouldn't we have it only once, at the very beginning? If we receive requested tx IMO it should be marked as received, no matter what happens next.

Right now in case of underpaid tx, we will request&reject it endlessly

File: src/Nethermind/Nethermind.TxPool/TxPool.cs [hunk 22]
```diff
@@ -1119,6 +1129,8 @@ private static void DisposeBlockAccountChanges(Block block)
         {
             block.DisposeAccountChanges();
         }
+
+        public AnnounceResult AnnounceTx(ValueHash256 txhash, IMessageHandler<PooledTransactionRequestMessage> retryHandler) => _retryCache.Announced(txhash, retryHandler);
```
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.TxPool/TxPool.cs]: Nitpick: would be good to move it a bit higher, to not be hidden in the codebase at the very end, after TxPoolReport contents

File: src/Nethermind/Nethermind.TxPool/TxPool.cs (hunk 21 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs]: Yeah, let's do this. I can't really think of a situation when mark only accepted ones is better

**Review by marcindsobczak — APPROVED** [review]
